File: 0051-n-queens/0051-n-queens.cpp
class Solution {
    unordered_set<int> rowPosition;
    unordered_set<int> posDiagonal;
    unordered_set<int> negDiagonal;

private:
    void backtrack(int col, int n, vector<string>& board,
                   vector<vector<string>>& ans) {
        if (col == n) {
            ans.push_back(board);
            return;
        }

        for (int row = 0; row < n; row++) {
            if (rowPosition.find(row) != rowPosition.end() || posDiagonal.find(row - col) != posDiagonal.end() ||
                negDiagonal.find(row + col) != negDiagonal.end()) {
                continue;
            }
            rowPosition.insert(row);
            posDiagonal.insert(row - col);
            negDiagonal.insert(row + col);
            board[row][col] = 'Q';

            backtrack(col + 1, n, board, ans);

            rowPosition.erase(row);
            posDiagonal.erase(row - col);
            negDiagonal.erase(row + col);
            board[row][col] = '.';
        }
    }

public:
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> ans;
        vector<string> board(n, string(n, '.'));
        backtrack(0, n, board, ans);
        return ans;
    }
};

File: 0051-n-queens/0051-n-queens.java
class Solution {
    private Set<Integer> cols = new HashSet<>();
    private Set<Integer> posDiag = new HashSet<>(); // r + c
    private Set<Integer> negDiag = new HashSet<>(); // r - c

    private void backtrack(int row, int n, List<String> board, List<List<String>> result) {
        if (row == n) {
            result.add(new ArrayList<>(board));
            return;
        }

        for (int col = 0; col < n; col++) {
            if (cols.contains(col) || posDiag.contains(row + col) || negDiag.contains(row - col)) {
                continue;
            }

            // Place the queen
            char[] rowArray = new char[n];
            Arrays.fill(rowArray, '.');
            rowArray[col] = 'Q';
            board.add(new String(rowArray));
            cols.add(col);
            posDiag.add(row + col);
            negDiag.add(row - col);

            // Move to the next row
            backtrack(row + 1, n, board, result);

            // Backtrack
            board.remove(board.size() - 1);
            cols.remove(col);
            posDiag.remove(row + col);
            negDiag.remove(row - col);
        }
    }

    public List<List<String>> solveNQueens(int n) {
        List<List<String>> result = new ArrayList<>();
        backtrack(0, n, new ArrayList<>(), result);
        return result;
    }
}